networkx/algorithms/euler.py

```diff
@@ -147,3 +158,231 @@ def eulerian_circuit(G, source=None):
             arbitrary_edge = next(edges(current_vertex))
             vertex_stack.append(get_vertex(arbitrary_edge))
             g.remove_edge(*arbitrary_edge)
+
+
+def has_eulerian_path(G):
+    """Returns ``True'' iff ``G'' has an Eulerian path.
+
+    An Eulerian path is a path that crosses every edge in G
+    exactly once.
+
+    Parameters
+    ----------
+    G: NetworkX Graph, DiGraph, MultiGraph or MultiDiGraph
+        A directed or undirected Graph or MultiGraph.
+
+    Returns
+    -------
+    True, False
+
+    Examples
+    --------
+    >>> G = nx.DiGraph([(1,2),(2,3)])                                      
+    >>> nx.has_eulerian_path(G)
+    True
+    """
+    is_directed = G.is_directed()
+
+    # Verify that graph is connected, short circuit
+    if is_directed and not nx.is_weakly_connected(G):
+        return False
+
+    # is undirected
+    if not is_directed and not nx.is_connected(G):
+        return False
+
+    # Now verify if has an Eulerian circuit: even condition of all
+    # nodes is satified.
+    if nx.is_eulerian(G):
```

Comment:
Actually, building the circuit is up to |V|^2. It's cheaper to run
is_connected() again.
On Dec 21, 2015 1:21 PM, "Humberto G. Ortiz Zuazaga" <humberto.ortiz@upr.edu>
wrote:

> I could try to build a circuit and catch any exceptions. It's still O(|V|).
> On Dec 21, 2015 12:03 PM, "Dan Schult" <notifications@github.com> wrote:
>
>> In networkx/algorithms/euler.py
>> <https://github.com/networkx/networkx/pull/1878#discussion_r48159556>:
>>
>> > +    >>> nx.has_eulerian_path(G)
>> > +    True
>> > +    """
>> > +    is_directed = G.is_directed()
>> > +
>> > +    # Verify that graph is connected, short circuit
>> > +    if is_directed and not nx.is_weakly_connected(G):
>> > +        return False
>> > +
>> > +    # is undirected
>> > +    if not is_directed and not nx.is_connected(G):
>> > +        return False
>> > +
>> > +    # Now verify if has an Eulerian circuit: even condition of all
>> > +    # nodes is satified.
>> > +    if nx.is_eulerian(G):
>>
>> Does this check nx.is_connected(G) a second time? Is there a way to avoid
>> duplicating the is_connected?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/networkx/networkx/pull/1878/files#r48159556>.
>>
>